### ISSUE TITLE
Global resource limits 

### ISSUE THREAD

**Issue by ozahavi** [aws/karpenter#2474]
**Tell us about your request**
The ability to configure resource limits (CPU\GPU) on a global level that will be enforced\applied to all provisioners in the cluster.

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
In certain use cases, I need to enable a mix of several CPU and GPU instance types that are configured in separate provisioners (because of different taints that are added to the nodes). Without a global limit for resources, I cannot limit the overall capacity of nodes created by Karpenter in the cluster.

**Are you currently working around this issue?**
No.

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by spring1843** [comment]
Thank you for the interesting suggestion.

Maybe you can use [ResourceQuotas](https://kubernetes.io/docs/concepts/policy/resource-quotas/) to achieve this? 

>Without a global limit for resources, I cannot limit the overall capacity of nodes created by Karpenter in the cluster.

Doesn't the union of all resource limits in all provisioners equal the global resource limits? e.g. If you have two provisioners with LimitA and limitB then your are limiting the global limits by (LimitA and LimitB). If yes then effectively you are able to limit the overall capacity of nodes created by Karpenter, just not in one place.

**Comment by ozahavi** [comment]
> Maybe you can use [ResourceQuotas](https://kubernetes.io/docs/concepts/policy/resource-quotas/) to achieve this?

ResourceQuotas are namespace bound, and I want the limit to be cross-namespaces (cluster level).
 
> Doesn't the union of all resource limits in all provisioners equal the global resource limits? e.g. If you have two provisioners with LimitA and limitB then your are limiting the global limits by (LimitA and LimitB). If yes then effectively you are able to limit the overall capacity of nodes created by Karpenter, just not in one place.

Not for my specific use case. I'll give an example. Let's say I have two teams that run workloads on an EKS cluster on multiple namespaces, and each team has a dedicated provisioner (each provisioner with different instance types). I have a budget of X for AWS spend, and I want to allow each team to spend X at any given moment potentially, but if I configure X as a limit for both provisioners, my spend can reach 2X. And that is why if I could provide a global limit, I would be able to control compute usage on a cluster level.

**Comment by spring1843** [comment]
>Not for my specific use case. I'll give an example. Let's say I have two teams that run workloads on an EKS cluster on multiple namespaces, and each team...

I see what you mean, you could use X/2 in this case then the sum would be X, but that doesn't give you what you are asking exactly since one cluster may end up using all of X/2 and the other only 10% of it. So a global limit would help here.

**Comment by stevehipwell** [comment]
We have this issue, as best practice for spot nodes is to opt in which requires at least two provisioners. In reality we have multiple provisioners to allow Karpenter to create tainted nodes as well as non-Karpenter nodes created from an ASG to run Karpenter on.

I'd also like to see Karpenter including nodes that it didn't provision in the calculation, it'd be fine if we had to label the nodes to make them part of this calculation.

This functionality is already available in Cluster Autoscaler so not having this makes migrating to Karpenter harder to sell.

**Comment by myaser** [comment]
we have a slightly different usecase that I requested in #4462 

we would like to have a global limit on the count of nodes or machines.

The limitation we have is on the node count (limited IP range available for the nodes), so it does not matter what global CPU and memory limit we define.

We can not enforce this using resource limits like CPU unless we are very strict about the instance-types. 

In fact, we would like not to limit the overall capacity of the cluster in terms of CPU and memory, allowing karpenter to replace smaller nodes with larger ones creating more CPU and memory without breaking the node limit

**Comment by njtran** [comment]
> we would like to have a global limit on the count of nodes or machines.

Yeah, if we do the global resource limits, similar to how we could extend the limits to be for custom resources, it would be equally possible to do this for the number of nodes/machines. Currently with limits, you can choose to do CPU and not memory I believe, and you could do the same by defining only a node limit instead of cpu/memory limits. 

This can also be extended for the Provisioner limits too or any other future references to a resource list throughout Karpenter to allow the same resource-types across all requests/limits fields where it makes sense.
